### ISSUE TITLE
Proposal: Add semgrep-go checks into validation step

### ISSUE THREAD

**Issue by benjaminch** [prebid/prebid-server#1551]
Discovered recently this amazing semgrep-go rules collection put together by @dgryski here https://github.com/dgryski/semgrep-go allowing to avoid common issue patterns in Go.

The check shouldn't by blocking for the time being, but once all errors / warnings will be solved, maybe it should become blocking in case something is detected.

I will be pushing a PR to introduce this one in the workflow so we can discuss and let me know what you think about it.

Cheers ! :)

**Comment by dlukeomalley** [comment]
Hey @benjaminch! Semgrep maintainer here. @dgryski's rules are also published to the Semgrep registry if you'd like to skip the `curl`:

```bash
$ semgrep --config="https://semgrep.dev/c/r/dgryski.semgrep-go"
```

This is the equivalent and you don't have to do the `.gitignore` etc. Either way works though and would to hear your feedback.

**Comment by benjaminch** [comment]
Hello @dlukeomalley !

Thanks a lot for your help here, really appreciate :)
I will try to use action directly, but first wanted to make this check part of the validation script so it's also available locally to contributors.

Let's see if I can make it happen !

Thanks !

**Comment by dlukeomalley** [comment]
Let me know if you run into any issues, the maintainer group loves supporting use cases like yours! 🙂

**Comment by SyntaxNode** [comment]
@benjaminch This looks very interesting. Thank you for taking the initiative.

I discovered this recently as well thanks to Grayson Hardaway's GoLang NYC talk. We're currently in the middle of revitalizing our CI/CD pipeline using GitHub Actions. This looks like a nice addition.

@mansinahar is working on adding a linter. I'm curious if you recommend this as a complimentary static analysis or a replacement to a traditional linter?

@dlukeomalley Thank you very much for the support. I'm excited to increase our code safety checks and also at the prospect of codifying our project specific rules to make PR reviews quicker. One use case I have in mind is catching writes to shared memory. Go lacks any sort of way to express read / write access, which is perhaps the thing I miss the most from C#. I imagine a rule of this kind would require flow analysis, right? Or is there a way to write a semgrep rule which catches a write to a variable which hasn't been copied first. Ex:

```
type Request struct {
    Site  *Site
    .. other stuff ...
}

type Site struct {
    Publisher string
}

BAD:
bidRequest.Site.Publisher = "foo"

GOOD:
siteCopy := *bidRequest.Site
siteCopy.Publisher = "foo" 
bidRequest.Site = &siteCopy
```


I'm not sure how to write a rule which prevents writing to a variable unless it's been copied first.

**Comment by benjaminch** [comment]
Hey @SyntaxNode !

> @mansinahar is working on adding a linter. I'm curious if you recommend this as a complimentary static analysis or a replacement to a traditional linter?

I think it's a complimentary static analysis tool along the linter, it won't replace a linter IMO as it won't serve the same purpose.

**Comment by mansinahar** [comment]
@SyntaxNode I agree with @benjaminch that this won't replace the linter as the linter checks more for style related rules vs this seems to check for potential bugs and such.

Also, I did find a few other static analysis tools for Go like [goreporter](https://github.com/360EntSecGroup-Skylar/goreporter), [go-critic](https://github.com/go-critic/go-critic) (This has it's own set of rules and plus also supports a `ruleguard` integration) and [staticcheck](https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck) which look interesting and useful too and so wanted to throw them in the mix here as well so that we can consider them all and pick the one suited best for [Prebid-Server](https://github.com/prebid/prebid-server).

I personally found [goreporter](https://github.com/360EntSecGroup-Skylar/goreporter) the most interesting as it supports a bunch of different linters and normalizes their output but I am not sure if that project is maintained any longer :(

**Comment by mansinahar** [comment]
@dlukeomalley out of curiosity, do you consider integrating [semgrep-go](https://github.com/dgryski/semgrep-go) with [golangci-lint](https://github.com/golangci/golangci-lint) in the future?

**Comment by dgryski** [comment]
My suggestion is to start by integrating tools with little to no false positives: go vet, staticcheck, ineffassign, and nilness. Most of the other liners catch note that are either more style issues and just generate code churn.

**Comment by dmitris** [comment]
for reference, here's the output of running semgrep with the https://github.com/dgryski/semgrep-go ruleset: 
https://gist.github.com/dmitris/908a85d205249f2018f4688308fb5053

I would be interested in fixing them and also adding this to the CI checks, if possible and reasonable.

**Comment by bsardo** [comment]
The PBS-Go engineering subcommittee is currently looking at ways to automate as much of the review process as possible getting authors timely feedback on best practices, etc prior to human review. We will revisit the ideas discussed here as part of that initiative.
